hacks/glx/kaleidocycle.man
.TH XScreenSaver 1 "25-Jul-98" "X Version 11"
.SH NAME
kaleidocycle \- draws twistable rings of tetrahedra
.SH SYNOPSIS
.B kaleidocycle
[\-\-display \fIhost:display.screen\fP] [\-\-window] [\-\-root]
[\-\-window\-id \fInumber\fP]
[\-\-visual \fIvisual\fP] [\-\-delay \fImicroseconds\fP]
[\-\-count \fInumber\fP]
[\-\-speed \fInumber\fP]
[\-\-wander] [\-\-no-wander]
[\-\-spin \fIaxes\fP]
[\-\-no-spin]
[\-\-wireframe]
[\-\-fps]
.SH DESCRIPTION
The \fIkaleidocycle\fP program draws a ring composed of tetrahedra
connected at the edges that twists and rotates toroidally.  Segments
are occasionally inserted or removed.
.SH OPTIONS
.I kaleidocycle
accepts the following options:
.TP 8
.B \-\-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-\-root
Draw on the root window.
.TP 8
.B \-\-window\-id \fInumber\fP
Draw on the specified window.
.TP 8
.B \-\-install
Install a private colormap for the window.
.TP 8
.B \-\-visual \fIvisual\fP\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-\-count \fInumber\fP
The initial number of segments. Default 16.  Must be 8 or greater, and
an even number.
.TP 8
.B \-\-speed \fInumber\fP
Adjust the animation speed. 0.5 for half as fast, 2.0 for twice as fast.
.TP 8
.B \-\-wander
Move the text around the screen.  This is the default.
.TP 8
.B \-\-no\-wander
Keep the text centered on the screen.
.TP 8
.B \-\-spin
Which axes around which the text should spin.  The default is "Z",
meaning rotate the object the plane of the screen only.
.TP 8
.B \-\-no\-spin
Don't spin the text at all: the same as \fB\-\-spin ""\fP.
.TP 8
.B \-\-wireframe
Render in wireframe instead of solid.
.TP 8
.B \-\-fps
Display the current frame rate, CPU load, and polygon count.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 8
.B XSCREENSAVER_WINDOW
The window ID to use with \fI\-\-root\fP.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1)
.SH COPYRIGHT
Copyright \(co 2001-2013 by Jamie Zawinski.
Permission to use, copy, modify, distribute, and sell this software and
its documentation for any purpose is hereby granted without fee,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation.  No representations are made about the
suitability of this software for any purpose.  It is provided "as is"
without express or implied warranty.
.SH AUTHOR
Jamie Zawinski <jwz@jwz.org>


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/glx/kaleidocycle.c
/* kaleidocycle, Copyright (c) 2013-2014 Jamie Zawinski <jwz@jwz.org>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 *
 * A loop of rotating tetrahedra.  Created by jwz, July 2013.
 * Inspired by, and some math borrowed from:
 * http://www.kaleidocycles.de/pdf/kaleidocycles_theory.pdf
 * http://intothecontinuum.tumblr.com/post/50873970770/an-even-number-of-at-least-8-regular-tetrahedra
 */


#define DEFAULTS	"*delay:	30000       \n" \
			"*count:        16          \n" \
			"*showFPS:      False       \n" \
			"*wireframe:    False       \n" \
			"*suppressRotationAnimation: True\n" \

# define release_kaleidocycle 0

#include "xlockmore.h"
#include "colors.h"
#include "normals.h"
#include "rotator.h"
#include "gltrackball.h"
#include <ctype.h>
#include <math.h>

#ifdef USE_GL /* whole file */

#define DEF_SPIN        "Z"
#define DEF_WANDER      "False"
#define DEF_SPEED       "1.0"

typedef struct {
  GLXContext *glx_context;
  rotator *rot, *rot2;
  trackball_state *trackball;
  Bool button_down_p;

  int min_count, max_count;
  Bool startup_p;

  int ncolors;
  XColor *colors;
  int ccolor;

  GLfloat count;
  GLfloat th, dth;

  enum { STATIC, IN, OUT } mode, prev_mode;

} kaleidocycle_configuration;

static kaleidocycle_configuration *bps = NULL;

static char *do_spin;
static GLfloat speed;
static Bool do_wander;

static XrmOptionDescRec opts[] = {
  { "-spin",   ".spin",   XrmoptionSepArg, 0 },
  { "+spin",   ".spin",   XrmoptionNoArg, "" },
  { "-wander", ".wander", XrmoptionNoArg, "True" },
  { "+wander", ".wander", XrmoptionNoArg, "False" },
  { "-speed",  ".speed",  XrmoptionSepArg, 0 },
};

static argtype vars[] = {
  {&do_spin,   "spin",   "Spin",   DEF_SPIN,   t_String},
  {&do_wander, "wander", "Wander", DEF_WANDER, t_Bool},
  {&speed,     "speed",  "Speed",  DEF_SPEED,  t_Float},
};

ENTRYPOINT ModeSpecOpt kaleidocycle_opts = {countof(opts), opts, countof(vars), vars, NULL};



/* Window management, etc
 */
ENTRYPOINT void
reshape_kaleidocycle (ModeInfo *mi, int width, int height)
{
  GLfloat h = (GLfloat) height / (GLfloat) width;
  int y = 0;

  if (width > height * 5) {   /* tiny window: show middle */
    height = width * 9/16;
    y = -height/2;
    h = height / (GLfloat) width;
  }

  glViewport (0, y, (GLint) width, (GLint) height);

  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  gluPerspective (30.0, 1/h, 1.0, 100.0);

  glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
  gluLookAt( 0.0, 0.0, 30.0,
             0.0, 0.0, 0.0,
             0.0, 1.0, 0.0);

  {
    GLfloat s = (MI_WIDTH(mi) < MI_HEIGHT(mi)
                 ? (MI_WIDTH(mi) / (GLfloat) MI_HEIGHT(mi))
                 : 1);
    glScalef (s, s, s);
  }

  glClear(GL_COLOR_BUFFER_BIT);
}


ENTRYPOINT Bool
kaleidocycle_handle_event (ModeInfo *mi, XEvent *event)
{
  kaleidocycle_configuration *bp = &bps[MI_SCREEN(mi)];

  if (gltrackball_event_handler (event, bp->trackball,
                                 MI_WIDTH (mi), MI_HEIGHT (mi),
                                 &bp->button_down_p))
    return True;
  else if (event->xany.type == KeyPress)
    {
      KeySym keysym;
      char c = 0;
      XLookupString (&event->xkey, &c, 1, &keysym, 0);
      if (c == '+' || c == '.' || c == '>' || c == '=' || c == '+' ||
          keysym == XK_Right || keysym == XK_Up || keysym == XK_Next)
        {
          bp->mode = IN;
          return True;
        }
      else if ((c == '-' || c == ',' || c == '<' || c == '-' || c == '_' ||
               keysym == XK_Left || keysym == XK_Down || keysym == XK_Prior) &&
               bp->count > bp->min_count)
        {
          bp->mode = OUT;
          return True;
        }
      else if (screenhack_event_helper (MI_DISPLAY(mi), MI_WINDOW(mi), event))
        goto DEF;
    }
  else if (screenhack_event_helper (MI_DISPLAY(mi), MI_WINDOW(mi), event))
    {
    DEF:
      if (bp->count <= bp->min_count)
        bp->mode = IN;
      else
        bp->mode = (random() & 1) ? IN : OUT;
      return True;
    }

  return False;
}



ENTRYPOINT void 
init_kaleidocycle (ModeInfo *mi)
{
  kaleidocycle_configuration *bp;
  int wire = MI_IS_WIREFRAME(mi);
  int i;

  MI_INIT (mi, bps);

  bp = &bps[MI_SCREEN(mi)];

  bp->glx_context = init_GL(mi);

  reshape_kaleidocycle (mi, MI_WIDTH(mi), MI_HEIGHT(mi));

  glLineWidth (4);

  if (!wire)
    {
      GLfloat pos[4] = {1.0, 1.0, 1.0, 0.0};
      GLfloat amb[4] = {0.0, 0.0, 0.0, 1.0};
      GLfloat dif[4] = {1.0, 1.0, 1.0, 1.0};
      GLfloat spc[4] = {0.0, 1.0, 1.0, 1.0};

      glEnable(GL_LIGHTING);
      glEnable(GL_LIGHT0);
      glEnable(GL_DEPTH_TEST);
      glEnable(GL_CULL_FACE);

      glLightfv(GL_LIGHT0, GL_POSITION, pos);
      glLightfv(GL_LIGHT0, GL_AMBIENT,  amb);
      glLightfv(GL_LIGHT0, GL_DIFFUSE,  dif);
      glLightfv(GL_LIGHT0, GL_SPECULAR, spc);

      glEnable (GL_BLEND);
      glBlendFunc (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    }

  {
    Bool spinx = False, spiny = False, spinz = False;
    double spin_speed   = 0.25;
    double wander_speed = 0.005;
    double spin_accel   = 0.2;
    double twist_speed  = 0.25;
    double twist_accel  = 1.0;

    char *s = do_spin;
    while (*s)
      {
        if      (*s == 'x' || *s == 'X') spinx = True;
        else if (*s == 'y' || *s == 'Y') spiny = True;
        else if (*s == 'z' || *s == 'Z') spinz = True;
        else if (*s == '0') ;
        else
          {
            fprintf (stderr,
         "%s: spin must contain only the characters X, Y, or Z (not \"%s\")\n",
                     progname, do_spin);
            exit (1);
          }
        s++;
      }

    bp->rot = make_rotator (spinx ? spin_speed : 0,
                            spiny ? spin_speed : 0,
                            spinz ? spin_speed : 0,
                            spin_accel,
                            do_wander ? wander_speed : 0,
                            False);
    bp->rot2 = make_rotator (twist_speed, 0, 0, twist_accel, 0, True);

    bp->trackball = gltrackball_init (True);
  }

  if (MI_COUNT(mi) < 8) MI_COUNT(mi) = 8;
  if (MI_COUNT(mi) & 1) MI_COUNT(mi)++;

  bp->min_count = 8;
  bp->max_count = 12 + MI_COUNT(mi) * 1.3;
  if (bp->max_count & 1) bp->max_count++;
  bp->startup_p = True;

  bp->count = 0;
  bp->mode = IN;
  bp->prev_mode = IN;

/*
  bp->count = MI_COUNT(mi);
  bp->mode = STATIC;
*/

  bp->ncolors = 512;
  if (! bp->colors)
    bp->colors = (XColor *) calloc(bp->ncolors, sizeof(XColor));
  make_uniform_colormap (0, 0, 0,
                         bp->colors, &bp->ncolors,
                         False, 0, False);

  for (i = 0; i < bp->ncolors; i++)
    {
      /* make colors twice as bright */
      bp->colors[i].red   = (bp->colors[i].red   >> 2) + 0x7FFF;
      bp->colors[i].green = (bp->colors[i].green >> 2) + 0x7FFF;
      bp->colors[i].blue  = (bp->colors[i].blue  >> 2) + 0x7FFF;
    }

  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
}



/* t = toroidal rotation, a = radial position
   colors = 4 colors, 4 channels each.
 */
static void
draw_tetra (ModeInfo *mi, double t, double a, Bool reflect_p,
            GLfloat *colors)
{
  int wire = MI_IS_WIREFRAME(mi);

  XYZ v1, v2, v3, P, Q;
  XYZ verts[4];
  int i;

  double scale;
  double sint = sin(t);
  double cost = cos(t);
  double tana = tan(a);
  double sint2 = sint * sint;
  double tana2 = tana * tana;

  v1.x = cost;
  v1.y = 0;
  v1.z = sint;

  scale = 1 / sqrt (1 + sint2 * tana2);
  v2.x = scale * -sint;
  v2.y = scale * -sint * tana;
  v2.z = scale * cost;

  v3.x = scale * -sint2 * tana;
  v3.y = scale;
  v3.z = scale * cost * sint * tana;

  P.x = v3.y / tana - v3.x;
  P.y = 0;
  P.z = -v3.z / 2;

  Q.x = v3.y / tana;
  Q.y = v3.y;
  Q.z = v3.z / 2;

  verts[0] = P;
  verts[1] = P;
  verts[2] = Q;
  verts[3] = Q;

  scale = sqrt(2) / 2;
  verts[0].x = P.x - scale * v1.x;
  verts[0].y = P.y - scale * v1.y;
  verts[0].z = P.z - scale * v1.z;

  verts[1].x = P.x + scale * v1.x;
  verts[1].y = P.y + scale * v1.y;
  verts[1].z = P.z + scale * v1.z;

  verts[2].x = Q.x - scale * v2.x;
  verts[2].y = Q.y - scale * v2.y;
  verts[2].z = Q.z - scale * v2.z;

  verts[3].x = Q.x + scale * v2.x;
  verts[3].y = Q.y + scale * v2.y;
  verts[3].z = Q.z + scale * v2.z;

  for (i = 0; i < 4; i++)
    {
      Bool reflect2_p = ((i + (reflect_p != 0)) & 1);
      XYZ a = verts[(i+1) % 4];
      XYZ b = verts[(i+2) % 4];
      XYZ c = verts[(i+3) % 4];
      XYZ n = ((i & 1)
               ? calc_normal (b, a, c)
               : calc_normal (a, b, c));
      if (wire)
        glColor4fv (colors + (i * 4));
      else
        glMaterialfv (GL_FRONT, GL_AMBIENT_AND_DIFFUSE, colors + (i * 4));

      glFrontFace (reflect2_p ? GL_CW : GL_CCW);
      glBegin (wire ? GL_LINE_LOOP : GL_TRIANGLES);
      glNormal3f (n.x, n.y, n.z);
      glVertex3f (a.x, a.y, a.z);
      glVertex3f (b.x, b.y, b.z);
      glVertex3f (c.x, c.y, c.z);
      glEnd();
    }
}


/* Reflect through the plane normal to the given vector.
 */
static void
reflect (double x, double y, double z)
{
  GLfloat m[4][4];

  m[0][0] = 1 - (2 * x * x);
  m[1][0] = -2 * x * y;
  m[2][0] = -2 * x * z;
  m[3][0] = 0;

  m[0][1] = -2 * x * y;
  m[1][1] = 1 - (2 * y * y);
  m[2][1] = -2 * y * z;
  m[3][1] = 0;

  m[0][2] = -2 * x * z;
  m[1][2] = -2 * y * z;
  m[2][2] = 1 - (2 * z * z);
  m[3][2] = 0;

  m[0][3] = 0;
  m[1][3] = 0;
  m[2][3] = 0;
  m[3][3] = 1;

  glMultMatrixf (&m[0][0]);
}


ENTRYPOINT void
draw_kaleidocycle (ModeInfo *mi)
{
  kaleidocycle_configuration *bp = &bps[MI_SCREEN(mi)];
  Display *dpy = MI_DISPLAY(mi);
  Window window = MI_WINDOW(mi);
  GLfloat colors[4*4];
  GLfloat count;
  double t, a;
  int i;

  GLfloat bspec[4] = {1.0, 1.0, 1.0, 1.0};
  GLfloat bshiny   = 128.0;

  if (!bp->glx_context)
    return;

  glXMakeCurrent(MI_DISPLAY(mi), MI_WINDOW(mi), *bp->glx_context);

  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

  mi->polygon_count = 0;

  glShadeModel(GL_SMOOTH);
  glEnable(GL_DEPTH_TEST);
  glEnable(GL_NORMALIZE);
  glEnable(GL_CULL_FACE);

  glPushMatrix ();

  {
    double x, y, z;
    get_position (bp->rot, &x, &y, &z, !bp->button_down_p);
    glTranslatef((x - 0.5) * 5,
                 (y - 0.5) * 5,
                 (z - 0.5) * 10);

    gltrackball_rotate (bp->trackball);

    get_rotation (bp->rot, &x, &y, &z, !bp->button_down_p);
    glRotatef (x * 360, 1, 0, 0);
    glRotatef (y * 360, 0, 1, 0);
    glRotatef (z * 360, 0, 0, 1);

    get_rotation (bp->rot2, &x, &y, &z, !bp->button_down_p);
    bp->th = x * 360 * 10 * speed;

    /* Make sure the twist is always in motion.  Without this, the rotator
       sometimes stops, and for too long, and it's boring looking.
     */
    bp->th += speed * bp->dth++;
    while (bp->dth > 360) bp->dth -= 360;
    while (bp->th  > 360) bp->th  -= 360;
  }

  glMaterialfv (GL_FRONT, GL_SPECULAR,  bspec);
  glMateriali  (GL_FRONT, GL_SHININESS, bshiny);


  /* Evenly spread the colors of the faces, and cycle them together.
   */
  for (i = 0; i < 4; i++)
    {
      int o  = bp->ncolors / 4;
      int c = (bp->ccolor + (o*i)) % bp->ncolors;
      colors[i*4+0] = bp->colors[c].red   / 65536.0;
      colors[i*4+1] = bp->colors[c].green / 65536.0;
      colors[i*4+2] = bp->colors[c].blue  / 65536.0;
      colors[i*4+3] = 1;
    }
  bp->ccolor++;
  if (bp->ccolor >= bp->ncolors) bp->ccolor = 0;


  count = (int) floor (bp->count);
  while (count < 8) count++;
  if (((int) floor (count)) & 1) count++;

  a = 2 * M_PI / (bp->count < 8 ? 8 : bp->count);
  t = bp->th / (180 / M_PI);

  glScalef (3, 3, 3);
  glScalef (a, a, a);
  glRotatef (90, 0, 0, 1);
/*  glRotatef (45, 0, 1, 0); */

  for (i = 0; i <= (int) floor (bp->count); i++)
    {
      Bool flip_p = (i & 1);
      glPushMatrix();
      glRotatef ((i/2) * 4 * 180 / bp->count, 0, 0, 1);
      if (flip_p) reflect (-sin(a), cos(a), 0);

      if (bp->mode != STATIC && i >= (int) floor (bp->count))
        {
          /* Fractional bp->count means the last piece is in transition */
          GLfloat scale = bp->count - (int) floor (bp->count);
          GLfloat tick = 0.07 * speed;
          GLfloat ocount = bp->count;
          GLfloat alpha;

          /* Fill in faster if we're starting up */
          if (bp->count < MI_COUNT(mi))
            tick *= 2;

          glScalef (scale, scale, scale);

          switch (bp->mode) {
          case IN: break;
          case OUT: tick = -tick; break;
          case STATIC: tick = 0; break;
          }

          bp->count += tick;

          if (bp->mode == IN
              ? floor (ocount) != floor (bp->count)
              : ceil  (ocount) != ceil  (bp->count))
            {
              if (bp->mode == IN)
                bp->count = floor (ocount) + 1;
              else
                bp->count = ceil  (ocount) - 1;

              if (((int) floor (bp->count)) & 1 ||
                  (bp->mode == IN && 
                   (bp->count < MI_COUNT(mi) &&
                    bp->startup_p)))
                {
                  /* keep going if it's odd, or less than 8. */
                  bp->count = round(bp->count);
                }
              else
                {
                  bp->mode = STATIC;
                  bp->startup_p = False;
                }
            }

          alpha = (scale * scale * scale * scale);
          if (alpha < 0.4) alpha = 0.4;
          colors[3] = colors[7] = colors[11] = colors[15] = alpha;
        }

      draw_tetra (mi, t, a, flip_p, colors);
      mi->polygon_count += 4;

      glPopMatrix();
    }

  if (bp->mode == STATIC && !(random() % 200)) {
    if (bp->count <= bp->min_count)
      bp->mode = IN;
    else if (bp->count >= bp->max_count)
      bp->mode = OUT;
    else
      bp->mode = bp->prev_mode;

    bp->prev_mode = bp->mode;
  }


  mi->recursion_depth = ceil (bp->count);

  glPopMatrix ();

  if (mi->fps_p) do_fps (mi);
  glFinish();

  glXSwapBuffers(dpy, window);
}


ENTRYPOINT void
free_kaleidocycle (ModeInfo *mi)
{
  kaleidocycle_configuration *bp = &bps[MI_SCREEN(mi)];
  if (!bp->glx_context) return;
  glXMakeCurrent(MI_DISPLAY(mi), MI_WINDOW(mi), *bp->glx_context);
  if (bp->trackball) gltrackball_free (bp->trackball);
  if (bp->rot) free_rotator (bp->rot);
  if (bp->rot2) free_rotator (bp->rot2);
  if (bp->colors) free (bp->colors);
}

XSCREENSAVER_MODULE ("Kaleidocycle", kaleidocycle)

#endif /* USE_GL */
